From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Tue, 19 May 2020 13:45:05 +0200
Subject: adds native audio processing

This patch adds custom denoising for around using a custom rnnoise model with support to set different NS levels and also adds events and handlers to process vad state and other audio related states and params inside libwebrtcs audio processing module.

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/media/engine/webrtc_voice_engine.cc b/media/engine/webrtc_voice_engine.cc
index 448ad3522a0d5562dfcd47b3d4db37bf50a708ae..18a67457bde24449df9f368b87b00df2bf485ef7 100644
--- a/media/engine/webrtc_voice_engine.cc
+++ b/media/engine/webrtc_voice_engine.cc
@@ -636,8 +636,7 @@ bool WebRtcVoiceEngine::ApplyOptions(const AudioOptions& options_in) {
   if (options.noise_suppression) {
     const bool enabled = *options.noise_suppression;
     apm_config.noise_suppression.enabled = enabled;
-    apm_config.noise_suppression.level =
-        webrtc::AudioProcessing::Config::NoiseSuppression::Level::kHigh;
+    // Allow audio_processing default for aggressiveness
     RTC_LOG(LS_INFO) << "NS set to " << enabled;
   }
 
diff --git a/modules/audio_processing/BUILD.gn b/modules/audio_processing/BUILD.gn
index d6ed3e251ba9bb546109a16a9a18cb8fe90e8b99..803411934979300f0c1dbad59130aadf542793e1 100644
--- a/modules/audio_processing/BUILD.gn
+++ b/modules/audio_processing/BUILD.gn
@@ -38,6 +38,7 @@ rtc_library("api") {
     "../../rtc_base/system:file_wrapper",
     "../../rtc_base/system:rtc_export",
     "agc:gain_control_interface",
+    "around_denoise",
   ]
   absl_deps = [ "//third_party/abseil-cpp/absl/types:optional" ]
 }
@@ -189,6 +190,9 @@ rtc_library("audio_processing") {
     "ns",
     "transient:transient_suppressor_api",
     "vad",
+    "around_biquad:around_biquad",
+    "around_biquad:mask_compensation_coefficients",
+    "around_agc:around_agc",
   ]
   absl_deps = [ "//third_party/abseil-cpp/absl/types:optional" ]
 
diff --git a/modules/audio_processing/aec3/echo_canceller3.cc b/modules/audio_processing/aec3/echo_canceller3.cc
index dfa932f49ce08ce2974dc5531470ce266ffe59a5..57e3f54206567480fa40a43ab3c242a118bcd9cb 100644
--- a/modules/audio_processing/aec3/echo_canceller3.cc
+++ b/modules/audio_processing/aec3/echo_canceller3.cc
@@ -85,7 +85,7 @@ void FillSubFrameView(
   RTC_DCHECK_GE(1, sub_frame_index);
   RTC_DCHECK_LE(0, sub_frame_index);
   RTC_DCHECK_EQ(frame->num_bands(), sub_frame_view->size());
-  RTC_DCHECK_EQ(frame->num_channels(), (*sub_frame_view)[0].size());
+  RTC_DCHECK_GE(frame->num_channels(), (*sub_frame_view)[0].size());
   for (size_t band = 0; band < sub_frame_view->size(); ++band) {
     for (size_t channel = 0; channel < (*sub_frame_view)[0].size(); ++channel) {
       (*sub_frame_view)[band][channel] = rtc::ArrayView<float>(
@@ -643,7 +643,7 @@ EchoCanceller3::RenderWriter::~RenderWriter() = default;
 void EchoCanceller3::RenderWriter::Insert(const AudioBuffer& input) {
   RTC_DCHECK_EQ(AudioBuffer::kSplitBandSize, input.num_frames_per_band());
   RTC_DCHECK_EQ(num_bands_, input.num_bands());
-  RTC_DCHECK_EQ(num_channels_, input.num_channels());
+  RTC_DCHECK_LE(num_channels_, input.num_channels());
 
   // TODO(bugs.webrtc.org/8759) Temporary work-around.
   if (num_bands_ != input.num_bands())
@@ -757,7 +757,7 @@ EchoCanceller3::~EchoCanceller3() = default;
 void EchoCanceller3::AnalyzeRender(const AudioBuffer& render) {
   RTC_DCHECK_RUNS_SERIALIZED(&render_race_checker_);
 
-  RTC_DCHECK_EQ(render.num_channels(), num_render_channels_);
+  RTC_DCHECK_GE(render.num_channels(), num_render_channels_);
   data_dumper_->DumpRaw("aec3_call_order",
                         static_cast<int>(EchoCanceller3ApiCall::kRender));
 
@@ -769,7 +769,7 @@ void EchoCanceller3::AnalyzeCapture(const AudioBuffer& capture) {
   data_dumper_->DumpWav("aec3_capture_analyze_input", capture.num_frames(),
                         capture.channels_const()[0], sample_rate_hz_, 1);
   saturated_microphone_signal_ = false;
-  for (size_t channel = 0; channel < capture.num_channels(); ++channel) {
+  for (size_t channel = 0; channel < num_capture_channels_; ++channel) {
     saturated_microphone_signal_ |=
         DetectSaturation(rtc::ArrayView<const float>(
             capture.channels_const()[channel], capture.num_frames()));
@@ -790,7 +790,7 @@ void EchoCanceller3::ProcessCapture(AudioBuffer* capture,
   RTC_DCHECK(capture);
   RTC_DCHECK_EQ(num_bands_, capture->num_bands());
   RTC_DCHECK_EQ(AudioBuffer::kSplitBandSize, capture->num_frames_per_band());
-  RTC_DCHECK_EQ(capture->num_channels(), num_capture_channels_);
+  RTC_DCHECK_GE(capture->num_channels(), num_capture_channels_);
   data_dumper_->DumpRaw("aec3_call_order",
                         static_cast<int>(EchoCanceller3ApiCall::kCapture));
 
diff --git a/modules/audio_processing/around_agc/BUILD.gn b/modules/audio_processing/around_agc/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..fb070f868e4252f64990e51b2e4d8d51cba7aca6
--- /dev/null
+++ b/modules/audio_processing/around_agc/BUILD.gn
@@ -0,0 +1,15 @@
+import("../../../webrtc.gni")
+
+rtc_source_set("around_agc") {
+  visibility = [ "../*" ]
+  configs += [ "..:apm_debug_dump" ]
+  sources = [
+    "around_agc.h",
+    "around_agc.cc",
+  ]
+  deps = [
+    "../agc2:noise_level_estimator",
+    "../agc2:adaptive_digital",
+    "../agc2:gain_applier",
+  ]
+}
diff --git a/modules/audio_processing/around_agc/around_agc.cc b/modules/audio_processing/around_agc/around_agc.cc
new file mode 100644
index 0000000000000000000000000000000000000000..4301a5efb2748a4ea09f90277f71c83c5d1f6666
--- /dev/null
+++ b/modules/audio_processing/around_agc/around_agc.cc
@@ -0,0 +1,101 @@
+#include "./around_agc.h"
+
+#include "common_audio/include/audio_util.h"
+#include "modules/audio_processing/audio_buffer.h"
+#include "modules/audio_processing/include/audio_frame_view.h"
+#include "modules/audio_processing/logging/apm_data_dumper.h"
+#include "rtc_base/atomic_ops.h"
+#include "rtc_base/checks.h"
+
+namespace webrtc {
+
+int AroundAgc::instance_count_ = 0;
+
+constexpr int kSpeechEstimatorAdjacentFramesThreshold = 1;
+constexpr float kSpeechEstimatorInitialSaturationMarginDb = 10.0f;
+constexpr float kSpeechEstimatorExtraSaturationMarginDb = 5.0f;
+constexpr int kGainApplierAdjacentSpeechFramesThreshold = 1;
+constexpr float kMaxGainChangePerSecondDb = 3.0f;
+constexpr float kFakeNoiseLevelDbfs = -50.0f;
+constexpr float kMaxBoostGainDb = 15.0f;
+
+AroundAgc::AroundAgc()
+    : data_dumper_(
+          new ApmDataDumper(rtc::AtomicOps::Increment(&instance_count_))),
+      adaptive_agc_(data_dumper_.get()),
+      limiter_(static_cast<size_t>(48000), data_dumper_.get(), "AroundAgc") {
+}
+
+AroundAgc::~AroundAgc() = default;
+
+void AroundAgc::Initialize(int sample_rate_hz) {
+  RTC_DCHECK(sample_rate_hz == AudioProcessing::kSampleRate8kHz ||
+             sample_rate_hz == AudioProcessing::kSampleRate16kHz ||
+             sample_rate_hz == AudioProcessing::kSampleRate32kHz ||
+             sample_rate_hz == AudioProcessing::kSampleRate48kHz);
+  limiter_.SetSampleRate(sample_rate_hz);
+}
+
+void AroundAgc::Process(AudioBuffer* audio, const float& vad_probability) {
+  AudioFrameView<float> frame(audio->channels(),
+                              audio->num_channels(),
+                              audio->num_frames());
+
+  // Run the adaptive AGC first
+  AdaptiveDigitalGainApplier::FrameInfo info;
+  float limiter_envelope = limiter_.LastAudioLevel();
+  float peak = 0.f;
+  float rms = 0.f;
+  for (const auto& x : frame.channel(0)) {
+    peak = std::max(std::fabs(x), peak);
+    rms += x * x;
+  }
+  VadLevelAnalyzer::Result vad_result = {vad_probability,
+                    FloatS16ToDbfs(std::sqrt(rms / frame.samples_per_channel())),
+                    FloatS16ToDbfs(peak)};
+  info.speech_probability = vad_result.speech_probability;
+  adaptive_agc_.speech_level_estimator_.Update(vad_result);
+  info.speech_level_dbfs = adaptive_agc_.speech_level_estimator_.level_dbfs();
+  info.speech_level_reliable = adaptive_agc_.speech_level_estimator_.IsConfident();
+  info.noise_rms_dbfs = kFakeNoiseLevelDbfs; // Set to fake noise level for max boost limiting.
+  info.headroom_db = adaptive_agc_.saturation_protector_->HeadroomDb();
+  info.limiter_envelope_dbfs =
+      limiter_envelope > 0.0f ? FloatS16ToDbfs(limiter_envelope) : -90.f;
+  adaptive_agc_.gain_applier_.Process(info, frame);
+  adaptive_agc_.saturation_protector_->Analyze(info.speech_probability, vad_result.peak_dbfs,
+                                 info.speech_level_dbfs);
+
+  // Then the limiter
+  limiter_.Process(frame);
+
+  // Finally, scale everything slightly to ensure headroom for encoding
+  constexpr float HEADROOM_SCALE = 0.94406f; // -0.5dB
+  for (auto& x : frame.channel(0)) {
+    x *= HEADROOM_SCALE;
+  }
+}
+
+void AroundAgc::Reset() {
+  adaptive_agc_.speech_level_estimator_.Reset();
+  limiter_.Reset();
+}
+
+AroundAgc::AdaptiveAgc::AdaptiveAgc(ApmDataDumper* apm_data_dumper)
+    : speech_level_estimator_(apm_data_dumper,
+                              kSpeechEstimatorAdjacentFramesThreshold),
+      gain_applier_(apm_data_dumper,
+                    kGainApplierAdjacentSpeechFramesThreshold,
+                    kMaxGainChangePerSecondDb,
+                    kFakeNoiseLevelDbfs + kMaxBoostGainDb,
+                    /* dry_run */ false),
+      noise_level_estimator_(CreateStationaryNoiseEstimator(apm_data_dumper)),
+      saturation_protector_(
+          CreateSaturationProtector(kSpeechEstimatorInitialSaturationMarginDb,
+                                    kSpeechEstimatorExtraSaturationMarginDb,
+                                    kSpeechEstimatorAdjacentFramesThreshold,
+                                    apm_data_dumper)) {
+}
+
+AroundAgc::AdaptiveAgc::~AdaptiveAgc() = default;
+
+}
diff --git a/modules/audio_processing/around_agc/around_agc.h b/modules/audio_processing/around_agc/around_agc.h
new file mode 100644
index 0000000000000000000000000000000000000000..29001e54b84d658891b26bc443e2ce5d0c61c8bf
--- /dev/null
+++ b/modules/audio_processing/around_agc/around_agc.h
@@ -0,0 +1,52 @@
+#ifndef MODULES_AUDIO_PROCESSING_AROUND_AGC_H_
+#define MODULES_AUDIO_PROCESSING_AROUND_AGC_H_
+
+#include <memory>
+#include <string>
+
+#include "modules/audio_processing/agc2/adaptive_digital_gain_applier.h"
+#include "modules/audio_processing/agc2/adaptive_mode_level_estimator.h"
+#include "modules/audio_processing/agc2/limiter.h"
+#include "modules/audio_processing/agc2/noise_level_estimator.h"
+#include "modules/audio_processing/agc2/saturation_protector.h"
+#include "modules/audio_processing/include/audio_processing.h"
+#include "modules/audio_processing/include/audio_frame_view.h"
+#include "rtc_base/constructor_magic.h"
+
+namespace webrtc {
+
+class ApmDataDumper;
+class AudioBuffer;
+
+// Applies adaptive digital gain control to a target speech level using
+// an external VAD signal. Based on AGC2::AdaptiveAgc.
+class AroundAgc {
+  public:
+  AroundAgc();
+  ~AroundAgc();
+  void Initialize(int sample_rate_hz);
+  void Process(AudioBuffer* audio, const float& vad_probability);
+  void Reset();
+
+ private:
+  struct AdaptiveAgc {
+    AdaptiveAgc(ApmDataDumper* apm_data_dumper);
+    ~AdaptiveAgc();
+
+    AdaptiveModeLevelEstimator speech_level_estimator_;
+    AdaptiveDigitalGainApplier gain_applier_;
+    std::unique_ptr<NoiseLevelEstimator> noise_level_estimator_;
+    std::unique_ptr<SaturationProtector> saturation_protector_;
+  };
+
+  static int instance_count_;
+  std::unique_ptr<ApmDataDumper> data_dumper_;
+  AdaptiveAgc adaptive_agc_;
+  Limiter limiter_;
+
+  RTC_DISALLOW_COPY_AND_ASSIGN(AroundAgc);
+};
+
+}
+
+#endif
diff --git a/modules/audio_processing/around_biquad/BUILD.gn b/modules/audio_processing/around_biquad/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..b0ec1cdf1b5f22e02756344aff88a8d45fbe3498
--- /dev/null
+++ b/modules/audio_processing/around_biquad/BUILD.gn
@@ -0,0 +1,21 @@
+import("../../../webrtc.gni")
+
+rtc_source_set("around_biquad") {
+  visibility = [ "../*" ]
+  configs += [ "..:apm_debug_dump" ]
+  sources = [
+    "around_biquad.h",
+    "around_biquad.cc",
+  ]
+  deps = [
+    "..:audio_buffer",
+  ]
+}
+
+rtc_source_set("mask_compensation_coefficients") {
+  visibility = [ "../*" ]
+  sources = [
+    "mask_compensation_coefficients.h",
+  ]
+  deps = []
+}
diff --git a/modules/audio_processing/around_biquad/around_biquad.cc b/modules/audio_processing/around_biquad/around_biquad.cc
new file mode 100644
index 0000000000000000000000000000000000000000..4fafad7e634cafc742c1cab61049eddc2450e0fc
--- /dev/null
+++ b/modules/audio_processing/around_biquad/around_biquad.cc
@@ -0,0 +1,39 @@
+#include "./around_biquad.h"
+#include "modules/audio_processing/include/audio_frame_view.h"
+
+namespace webrtc {
+
+static const float DEFAULT_COEFFICIENTS[] = {1.0f, 0.0f, 0.0f, 0.0f, 0.0f,};
+
+AroundBiquad::AroundBiquad() :
+  pCoefficients_(DEFAULT_COEFFICIENTS),
+  state_() {
+  Reset();
+}
+
+AroundBiquad::~AroundBiquad() = default;
+
+void AroundBiquad::Initialize(const float* pCoefficients) {
+  pCoefficients_ = pCoefficients;
+}
+
+void AroundBiquad::Reset() {
+  state_[0] = 0.0f;
+  state_[1] = 0.0f;
+}
+
+void AroundBiquad::Process(AudioBuffer* audio) {
+  AudioFrameView<float> frame(audio->channels(),
+                              audio->num_channels(),
+                              audio->num_frames());
+
+  float result;
+  for (float& x : frame.channel(0)) {
+    result = x * pCoefficients_[0] + state_[0];                                  // y[n] = x[n]*b0 + z[-1]
+    state_[0] = x * pCoefficients_[1] - result * pCoefficients_[3] + state_[1];  // z[-1] = x[n]*b1 - y[n]*a1 + z[-2]
+    state_[1] = x * pCoefficients_[2] - result * pCoefficients_[4];              // z[-2] = x[n]*b2 - y[n]*a2
+    x = result;
+  }
+}
+
+}
diff --git a/modules/audio_processing/around_biquad/around_biquad.h b/modules/audio_processing/around_biquad/around_biquad.h
new file mode 100644
index 0000000000000000000000000000000000000000..f4e1d16386526d9ac66644ffcaa9345997acb44f
--- /dev/null
+++ b/modules/audio_processing/around_biquad/around_biquad.h
@@ -0,0 +1,33 @@
+#ifndef MODULES_AUDIO_PROCESSING_AROUND_BIQUAD_H_
+#define MODULES_AUDIO_PROCESSING_AROUND_BIQUAD_H_
+
+#include "modules/audio_processing/audio_buffer.h"
+#include "rtc_base/constructor_magic.h"
+
+namespace webrtc {
+
+// Applies biquad processing to the first channel of the provided audio buffer.
+class AroundBiquad {
+  public:
+  AroundBiquad();
+  ~AroundBiquad();
+
+  // Sets the coefficients to use.
+  void Initialize(const float* pCoefficients);
+
+  // Resets internal state.
+  void Reset();
+
+  // Processes the first channel of the buffer.
+  void Process(AudioBuffer* audio);
+
+ private:
+  const float* pCoefficients_;
+  float state_[2];
+
+  RTC_DISALLOW_COPY_AND_ASSIGN(AroundBiquad);
+};
+
+}
+
+#endif
diff --git a/modules/audio_processing/around_biquad/mask_compensation_coefficients.h b/modules/audio_processing/around_biquad/mask_compensation_coefficients.h
new file mode 100644
index 0000000000000000000000000000000000000000..316d560ac8748787a2617b2b428a61ce5e14dc4f
--- /dev/null
+++ b/modules/audio_processing/around_biquad/mask_compensation_coefficients.h
@@ -0,0 +1,8 @@
+#ifndef MODULES_AUDIO_PROCESSING_MASK_COMPENSATION_COEFFICIENTS_H_
+#define MODULES_AUDIO_PROCESSING_MASK_COMPENSATION_COEFFICIENTS_H_
+
+// Biquad
+// +8dB High Shelf, f0 1400Hz, Q0.8, with -0.9dB fullband attenuation
+static const float MASK_COMPENSATION_COEFFS[] = {2.140794f, -3.883752f, 1.78456f, -1.704033f, 0.750257f,};
+
+#endif
diff --git a/modules/audio_processing/around_denoise/BUILD.gn b/modules/audio_processing/around_denoise/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..99ec07cdd04c740c668e2a119abd378b358eb4f5
--- /dev/null
+++ b/modules/audio_processing/around_denoise/BUILD.gn
@@ -0,0 +1,12 @@
+import("../../../webrtc.gni")
+
+rtc_source_set("around_denoise") {
+  visibility = [ "../*" ]
+  sources = [
+    "around_denoise.h",
+    "around_denoise.cc",
+  ]
+  deps = [
+    "//third_party/rnnoise_full:rnnoise_full_src",
+  ]
+}
diff --git a/modules/audio_processing/around_denoise/DEPS b/modules/audio_processing/around_denoise/DEPS
new file mode 100644
index 0000000000000000000000000000000000000000..a277a204037e9d9c3a586f42b026ded3705c213a
--- /dev/null
+++ b/modules/audio_processing/around_denoise/DEPS
@@ -0,0 +1,3 @@
+include_rules = [
+  "+third_party/rnnoise_full",
+]
diff --git a/modules/audio_processing/around_denoise/around_denoise.cc b/modules/audio_processing/around_denoise/around_denoise.cc
new file mode 100644
index 0000000000000000000000000000000000000000..d0142827b62757027b88fbf709dda1a4ebe4ab8f
--- /dev/null
+++ b/modules/audio_processing/around_denoise/around_denoise.cc
@@ -0,0 +1,474 @@
+#include <array>
+#include <math.h>
+
+#include "api/array_view.h"
+#include "common_audio/include/audio_util.h"
+#include "common_audio/window_generator.h"
+#include "modules/audio_processing/around_denoise/around_denoise.h"
+#include "modules/audio_processing/audio_buffer.h"
+#include "modules/audio_processing/include/audio_frame_view.h"
+#include "rtc_base/logging.h"
+#include "rtc_base/time_utils.h"
+
+namespace webrtc {
+
+#define DRR_ORDER 10
+#define DRR_COMMUNICATION_SLOWDOWN_THRESHOLD 4
+
+static const int VAD_DISABLE_TIMEOUT_MS = 1500;
+static const float VAD_DISABLE_THRESHOLD_PCT = 0.75;
+static const float VAD_ENABLED_THRESHOLD_PCT = 0.9;
+static const float VAD_SMOOTHING_WEMA_ALPHA = 0.09;
+static const float VAD_MUTED_SMOOTHING_WEMA_ALPHA = 0.05;
+static const size_t kDrrRingBufferSize = DRR_FRAME_SIZE * 2;
+
+float compute_wema(float prev_value, float new_data_point, float alpha) {
+  return prev_value > 0.0f
+    ? (1.0f - alpha) * prev_value + alpha * new_data_point
+    : new_data_point;
+}
+
+double wema_update(double alpha, double current_wema, double new_data_point) {
+  if (current_wema) {
+    double redistributed_mean = (1.0 - alpha) * current_wema;
+    double mean_increment = alpha * new_data_point;
+
+    return redistributed_mean + mean_increment;
+  } else {
+    return new_data_point;
+  }
+}
+
+double mean(double *signal, size_t signal_len) {
+  double sum = 0.0;
+  for (size_t i = 0; i < signal_len; ++i) {
+    sum += signal[i];
+  }
+  return sum / static_cast<double>(signal_len);
+}
+
+double normalize(double val, double min, double max) {
+  double res = (val - min) / (max - min);
+  return fmax(0.0, fmin(1.0, res));
+}
+
+double kurtosis(double *signal, size_t signal_len) {
+  double m = mean(signal, signal_len);
+  int n = static_cast<int>(signal_len);
+
+  double numerator = 0.0;
+  double denominator = 0.0;
+
+  for (int s = 0; s < n; ++s) {
+    double sample = signal[s];
+    double squared = (sample - m) * (sample - m);
+
+    denominator += squared;
+    numerator += squared * squared;
+  }
+
+  numerator = (1.0 / n) * numerator;
+  denominator = ((1.0 / n) * denominator) * ((1.0 / n) * denominator);
+
+  return numerator / denominator;
+}
+
+void filter(float *signal, size_t signal_len, double *ff, size_t ff_len, double fb, double *filtered) {
+  double inverse_fb = 1.0 / fb;
+
+  for (size_t sample = 0; sample < signal_len; ++sample) {
+    filtered[sample] = 0.0;
+
+    for (size_t ff_ix = 0; ff_ix < std::min(ff_len, sample + 1); ++ff_ix) {
+      filtered[sample] += ff[ff_ix] * signal[sample - ff_ix];
+    }
+
+    filtered[sample] *= inverse_fb;
+  }
+}
+
+void auto_correlation(const float* in_vector,
+                      size_t in_vector_length,
+                      size_t max_lags,
+                      double* result) {
+  double sum = 0.0;
+
+  for (size_t lag = 0; lag <= max_lags; ++lag) {
+    sum = 0;
+
+    for (size_t j = 0; j < in_vector_length - lag; ++j) {
+      sum += in_vector[j + lag] * in_vector[j];
+    }
+
+    result[lag] = sum;
+  }
+}
+
+void fix_autocorrelation(double *ac) {
+  // Dirty trick to ensure we don't get division by 0 in Levinson-Durbin. This
+  // should have minimal impact on the results.
+  if (ac[0] == 0.0f) {
+    ac[0] = ac[0] + fmax(0.5f * fabs(ac[1] - ac[0]),
+      std::numeric_limits<double>::epsilon());
+  }
+  // In the very unlikely event that epsilon makes the second coefficient equal
+  // to the first, we have to keep increasing the second coeff.
+  while (fabs(ac[0]) == fabs(ac[1])) {
+    ac[1] = ac[1] + fmax(0.5 * fabs(ac[1] - ac[0]),
+      std::numeric_limits<double>::epsilon());
+  }
+}
+
+double levinson_durbin(double *ac_coef, double *lpc_coef) {
+  double lpc_coef_temp[DRR_ORDER];
+  double lpc_error = ac_coef[0];
+
+  for (size_t i = 0; i < DRR_ORDER + 1; ++i) {
+    lpc_coef[i] = 0.0;
+  }
+
+  for (size_t i = 0; i < DRR_ORDER; ++i) {
+    lpc_coef_temp[i] = 0.0;
+  }
+
+  lpc_coef[0] = 1.0;
+  lpc_coef_temp[0] = 1.0;
+
+  for (size_t coeff_ix = 1; coeff_ix < DRR_ORDER + 1; ++coeff_ix) {
+    // Compute reflection coefficient.
+    double refl_coef = ac_coef[coeff_ix];
+
+    for (size_t ref_ix = 1; ref_ix < coeff_ix; ref_ix++) {
+      refl_coef += lpc_coef[ref_ix] * ac_coef[coeff_ix - ref_ix];
+    }
+
+    refl_coef /= lpc_error;
+
+    // Update coefficients.
+    lpc_coef[coeff_ix] = -1 * refl_coef;
+    for (size_t update_ix = 1; update_ix < coeff_ix; ++update_ix) {
+      lpc_coef_temp[update_ix] = lpc_coef[update_ix] -
+        refl_coef * lpc_coef[coeff_ix - update_ix];
+    }
+
+    for (size_t update_ix = 1; update_ix < coeff_ix; ++update_ix) {
+      lpc_coef[update_ix] = lpc_coef_temp[update_ix];
+    }
+
+    lpc_error *= 1 - refl_coef * refl_coef;
+  }
+
+  return fabs(lpc_error);
+}
+
+void lpr(float *signal, size_t signal_len, double *filtered) {
+  double ac_coef[DRR_ORDER + 2];
+
+  auto_correlation(signal, signal_len, DRR_ORDER + 1, ac_coef);
+
+  fix_autocorrelation(ac_coef);
+
+  double lpc_coef[DRR_ORDER + 1];
+  double lpc_error = levinson_durbin(ac_coef, lpc_coef);
+
+  filter(signal, signal_len, lpc_coef, DRR_ORDER + 1, sqrt(lpc_error), filtered);
+}
+
+AroundDenoiser::AroundDenoiser() = default;
+AroundDenoiser::~AroundDenoiser() {
+  rnnoise_destroy(denoise_state);
+}
+
+void AroundDenoiser::Initialize() {
+  RTC_LOG(LS_INFO) << "@AROUND: initialising denoiser";
+  denoise_state = rnnoise_create(NULL);
+  drr_buffer_ = WebRtc_CreateBuffer(kDrrRingBufferSize, sizeof(float));
+  running_avg_rms_buffer_ = WebRtc_CreateBuffer(settings_.averaging_window, sizeof(float));
+  running_avg_res_buffer_ = WebRtc_CreateBuffer(settings_.averaging_window, sizeof(double));
+  prev_kurtosis_ = 0.0;
+  drr_results_ = (double*)malloc(DRR_COMMUNICATION_SLOWDOWN_THRESHOLD * sizeof(double));
+  drr_results_index_ = 0;
+  WindowGenerator::Hanning(DRR_FRAME_SIZE, hanning_window_);
+}
+
+void AroundDenoiser::Denoise(AudioBuffer* audioBuffer, float noise_suppression_db) {
+  if (audioBuffer->num_channels() < 1) return;
+
+  AudioFrameView<float> frame(audioBuffer->channels_f(),
+                              audioBuffer->num_channels(),
+                              audioBuffer->num_frames());
+  // ensure the resamplers are set up correctly
+  int source_sample_rate = static_cast<int>(frame.samples_per_channel() * 100);
+  float* raw_frame_data = frame.channel(0).data();
+  InitializeResamplersIfNeeded(source_sample_rate);
+
+  std::array<float, kRNNoiseFrameSize> work_frame;
+  float* frame_data = work_frame.data();
+
+  // Feed the 1st channel to the resampler.
+  resampler_.Resample(raw_frame_data, frame.samples_per_channel(), frame_data,
+                      kRNNoiseFrameSize);
+
+  audio_features.vad_probability =
+      rnnoise_process_frame(denoise_state, frame_data, frame_data, noise_suppression_db);
+
+  ProcessVADState(audio_features.vad_probability, rtc::TimeUTCMillis());
+
+  // resample back to the rendering sample rate,
+  // writing the processed buffer into the output frame
+  reverse_resampler_.Resample(const_cast<const float*>(frame_data),
+                              kRNNoiseFrameSize, raw_frame_data,
+                              frame.samples_per_channel());
+}
+
+AroundDenoiser::AudioFeatures* AroundDenoiser::getCurrentFeatures() {
+  return &audio_features;
+}
+
+void AroundDenoiser::InitializeResamplersIfNeeded(int sample_rate_hz) {
+  resampler_.InitializeIfNeeded(sample_rate_hz, kRNNoiseSampleRate, 1);
+  reverse_resampler_.InitializeIfNeeded(kRNNoiseSampleRate, sample_rate_hz, 1);
+}
+
+void AroundDenoiser::CalculateRMS(AudioBuffer* audioBuffer) {
+  if (audioBuffer->num_channels() < 1) return;
+
+  AudioFrameView<float> frame(audioBuffer->channels_f(),
+                              audioBuffer->num_channels(),
+                              audioBuffer->num_frames());
+  rtc::ArrayView<float> signal = frame.channel(0);
+
+  float sum = 0.0;
+  for (auto& sample : signal) {
+    sum += pow(sample, 2);
+  }
+  audio_features.rms = sqrt(sum / signal.size());
+}
+
+float AroundDenoiser::ProcessDRREstimation(AudioBuffer* audioBuffer) {
+  AudioFrameView<float> frame(audioBuffer->channels_f(),
+                              audioBuffer->num_channels(),
+                              audioBuffer->num_frames());
+  float drr_to_emit = 0.0f;
+
+  if (is_microphone_on_ && frame.num_channels() > 0) {
+    rtc::ArrayView<float> signal = frame.channel(0);
+    float* signal_data = signal.data();
+    float analysis_data[DRR_FRAME_SIZE];
+    float vad = audio_features.vad_probability;
+
+    WebRtc_WriteBuffer(drr_buffer_, signal_data, signal.size());
+
+    size_t buffered_frame_length = WebRtc_available_read(drr_buffer_) / DRR_FRAME_SIZE;
+
+    while (buffered_frame_length > 0) {
+      if (is_in_group_ && drr_enabled_ && vad > VAD_DISABLE_THRESHOLD_PCT) {
+        WebRtc_ReadBuffer(drr_buffer_, NULL, analysis_data, DRR_FRAME_SIZE);
+        WebRtc_MoveReadPtr(drr_buffer_, -(DRR_FRAME_SIZE / 2));
+
+        float rms = FloatS16ToFloat(audio_features.rms);
+        double drr = FullPipelineV2Normalized(analysis_data, DRR_FRAME_SIZE, rms);
+
+        if (drr_results_index_ < DRR_COMMUNICATION_SLOWDOWN_THRESHOLD) {
+          drr_results_[drr_results_index_++] = drr;
+        }
+      } else {
+        WebRtc_MoveReadPtr(drr_buffer_, DRR_FRAME_SIZE);
+      }
+      buffered_frame_length = WebRtc_available_read(drr_buffer_) / DRR_FRAME_SIZE;
+    }
+  }
+
+  if (drr_results_index_ == DRR_COMMUNICATION_SLOWDOWN_THRESHOLD) {
+    float sum = 0.0f;
+    for (size_t i = 0; i < DRR_COMMUNICATION_SLOWDOWN_THRESHOLD; ++i) {
+      sum += drr_results_[i];
+    }
+
+    drr_to_emit = sum / DRR_COMMUNICATION_SLOWDOWN_THRESHOLD;
+    drr_results_index_ = 0;
+  }
+
+  return drr_to_emit;
+}
+
+AroundDenoiser::FullPipeline2Result AroundDenoiser::FullPipelineV2(float *analysis_data, size_t analysis_data_len, float rms) {
+  for (size_t i = 0; i < analysis_data_len; ++i) {
+    analysis_data[i] = FloatS16ToFloat(analysis_data[i]) * hanning_window_[i];
+  }
+
+  double filtered[analysis_data_len];
+  lpr(analysis_data, analysis_data_len, filtered);
+
+  double residual_kurtosis = kurtosis(filtered, analysis_data_len);
+
+  if (residual_kurtosis > 0 && prev_kurtosis_ &&
+      (residual_kurtosis - prev_kurtosis_) > settings_.extractors_residual_kurtosis_peak_threshold) {
+    // Prevent some unwanted kurtosis peaks that tend to skew the results
+    residual_kurtosis = prev_kurtosis_;
+  }
+
+  return AroundDenoiser::FullPipeline2Result{ rms, residual_kurtosis };
+}
+
+void AroundDenoiser::ResetDRR() {
+  drr_wema_ = 0.0f;
+  WebRtc_InitBuffer(running_avg_rms_buffer_);
+  WebRtc_InitBuffer(running_avg_res_buffer_);
+}
+
+double AroundDenoiser::FullPipelineV2Normalized(float *analysis_data, size_t analysis_data_len, float rms) {
+  AroundDenoiser::FullPipeline2Result result = FullPipelineV2(analysis_data, analysis_data_len, rms);
+
+  WebRtc_WriteBuffer(running_avg_rms_buffer_, &result.rms, 1);
+  WebRtc_WriteBuffer(running_avg_res_buffer_, &result.residual_kurtosis, 1);
+
+  size_t rms_size = WebRtc_available_read(running_avg_rms_buffer_);
+  size_t res_size = WebRtc_available_read(running_avg_res_buffer_);
+
+  float rms_window[rms_size];
+  double res_window[res_size];
+  WebRtc_ReadBuffer(running_avg_rms_buffer_, NULL, rms_window, rms_size);
+  WebRtc_ReadBuffer(running_avg_res_buffer_, NULL, res_window, res_size);
+
+  WebRtc_MoveReadPtr(running_avg_rms_buffer_,
+    rms_size < settings_.averaging_window ? -rms_size : 1 - rms_size);
+  WebRtc_MoveReadPtr(running_avg_res_buffer_,
+    res_size < settings_.averaging_window ? -res_size : 1 - res_size);
+
+  float rms_sum = 0.0f;
+  for (size_t i = 0; i < rms_size; ++i) {
+    rms_sum += rms_window[i];
+  }
+
+  double res_sum = 0.0f;
+  for (size_t i = 0; i < res_size; ++i) {
+    res_sum += res_window[i];
+  }
+
+  double rms_avg = rms_sum / rms_size;
+  double res_avg = res_sum / res_size;
+  double rms_normalized = normalize(rms_avg, settings_.normalization_rms_low, settings_.normalization_rms_high);
+  double res_normalized = normalize(res_avg, settings_.residual_kurtosis_low, settings_.residual_kurtosis_high);
+
+  // Residual Kurtosis often returns a series of 0-s out of a sudden. Let's try
+  // to ignore these to reduce to saw-tooth effect on the resulting data
+  if (rms_normalized > 0.0) {
+    last_valid_residual_kurtosis_ = rms_normalized;
+    last_valid_residual_kurtosis_distance_ = 0;
+  } else if (last_valid_residual_kurtosis_ > 0.0) {
+    last_valid_residual_kurtosis_distance_++;
+    if (
+      last_valid_residual_kurtosis_distance_ <=
+      kLastValidResidualKurtosisMaxConsideredDistance
+    ) {
+      rms_normalized = last_valid_residual_kurtosis_;
+    } else {
+      last_valid_residual_kurtosis_ = 0.0;
+    }
+  }
+
+  double drr = fmax(0.0, fmin(1.0,
+    res_normalized * settings_.extractors_residual_kurtosis_weight +
+    rms_normalized * settings_.extractors_rms_weight
+  ));
+
+  drr_wema_ = wema_update(
+    drr > drr_wema_ ? settings_.wema_inc_alpha : settings_.wema_dec_alpha,
+    drr_wema_,
+    drr
+  );
+
+  return drr_wema_;
+}
+
+void AroundDenoiser::EncodeFeaturesToSignal(AudioBuffer* audioBuffer) {
+  AudioFrameView<float> frame(audioBuffer->channels_f(),
+                              audioBuffer->num_channels(),
+                              audioBuffer->num_frames());
+  if (frame.num_channels() > 1) {
+    rtc::ArrayView<float> signal = frame.channel(1);
+    signal[0] = audio_features.vad_probability * INT16_MAX;
+    signal[1] = audio_features.rms;
+    // zero out the remaining samples
+    for (size_t sample = 2; sample < signal.size(); sample++) {
+      signal[sample] = 0;
+    }
+  }
+}
+
+void AroundDenoiser::ProcessMutedSpeakingState(float vad, int64_t now) {
+  if (
+    (vad_state_ == MutedStarted) &&
+    (vad_wema_ == 0.0 || vad_wema_ <= VAD_DISABLE_THRESHOLD_PCT) &&
+    muted_vad_disable_timeout_ > 0 &&
+    (now - muted_vad_disable_timeout_ >= VAD_DISABLE_TIMEOUT_MS)
+  ) {
+    vad_state_ = MutedStopped;
+    muted_vad_disable_timeout_ = 0;
+  }
+
+  vad_wema_ = compute_wema(vad_wema_, vad, VAD_MUTED_SMOOTHING_WEMA_ALPHA);
+
+  if (vad_wema_ >= VAD_ENABLED_THRESHOLD_PCT) {
+    vad_state_ = MutedStarted;
+    muted_vad_disable_timeout_ = now;
+  }
+}
+
+void AroundDenoiser::ProcessVADState(float vad, int64_t now) {
+  if (!is_microphone_on_) {
+    ProcessMutedSpeakingState(vad, now);
+    return;
+  }
+
+  if (
+    (vad_state_ == Started || vad_state_ == Sustained) &&
+    (vad == 0.0 || vad <= VAD_DISABLE_THRESHOLD_PCT) &&
+    vad_disable_timeout_ > 0 &&
+    (now - vad_disable_timeout_ >= VAD_DISABLE_TIMEOUT_MS)
+  ) {
+    vad_state_ = Stopped;
+    vad_disable_timeout_ = 0;
+  }
+
+  if (vad >= VAD_ENABLED_THRESHOLD_PCT) {
+    if (vad_state_ == Stopped || vad_state_ == MutedStarted || vad_state_ == MutedStopped) {
+      vad_state_ = Started;
+    }
+    vad_disable_timeout_ = now;
+  }
+
+  vad_wema_ = compute_wema(vad_wema_, vad, VAD_SMOOTHING_WEMA_ALPHA);
+
+  if (
+    vad_state_ == Started &&
+    vad > 0.0 && vad_wema_ > VAD_ENABLED_THRESHOLD_PCT &&
+    vad_state_ != Sustained
+  ) {
+    vad_state_ = Sustained;
+  }
+}
+
+void AroundDenoiser::SetAudioStatus(std::map<std::string, std::string> params) {
+  bool is_microphone_on = params["is_microphone_on"] == "true";
+  bool is_in_group = params["is_in_group"] == "true";
+
+  if (is_microphone_on_ && !is_microphone_on) {
+    vad_wema_ = 0.0;
+  } else if (!is_microphone_on_ && is_microphone_on) {
+    if (vad_state_ == MutedStarted) {
+      vad_state_ = MutedStopped;
+      muted_vad_disable_timeout_ = 0;
+    }
+  }
+
+  is_microphone_on_ = is_microphone_on;
+  is_in_group_ = is_in_group;
+}
+
+void AroundDenoiser::SetDRREnabled(bool enabled) {
+  drr_enabled_ = enabled;
+}
+
+} // namespace webrtc
diff --git a/modules/audio_processing/around_denoise/around_denoise.h b/modules/audio_processing/around_denoise/around_denoise.h
new file mode 100644
index 0000000000000000000000000000000000000000..8fb7563319fbd6844dad20f4111805549b8187dc
--- /dev/null
+++ b/modules/audio_processing/around_denoise/around_denoise.h
@@ -0,0 +1,132 @@
+#ifndef MODULES_AUDIO_PROCESSING_AROUND_DENOISE_H_
+#define MODULES_AUDIO_PROCESSING_AROUND_DENOISE_H_
+
+#include "common_audio/resampler/include/push_resampler.h"
+#include "modules/audio_processing/include/audio_processing.h"
+#include "rtc_base/constructor_magic.h"
+
+extern "C" {
+#include "common_audio/ring_buffer.h"
+#include "third_party/rnnoise_full/include/rnnoise.h"
+}
+
+namespace webrtc {
+
+#define DRR_FRAME_SIZE 1024
+
+constexpr size_t kRNNoiseFrameSize = 480;
+constexpr size_t kRNNoiseSampleRate = 48000;
+constexpr float kAroundDefaultDenoiserVADThreshold = 0.5;
+constexpr float kLastValidResidualKurtosisMaxConsideredDistance = 10;
+
+// Custom Denoiser module which uses the full RNNoise model.
+class AroundDenoiser {
+ public:
+  struct AudioFeatures {
+    float vad_probability;
+    float rms;
+  };
+
+  struct FullPipeline2Result {
+    float rms;
+    double residual_kurtosis;
+  };
+
+  struct FeatureExtractorSettings {
+    size_t averaging_window;
+    float wema_inc_alpha;
+    float wema_dec_alpha;
+    float normalization_rms_low;
+    float normalization_rms_high;
+    float residual_kurtosis_low;
+    float residual_kurtosis_high;
+    float spectral_difference_skewness_low;
+    float spectral_difference_skewness_high;
+    bool spectral_difference_skewness_flip;
+    float extractors_residual_kurtosis_peak_threshold;
+    float extractors_residual_kurtosis_weight;
+    float extractors_rms_weight;
+  };
+
+  enum VADState { Started, Stopped, Sustained, MutedStarted, MutedStopped };
+
+  AroundDenoiser();
+  ~AroundDenoiser();
+
+  void Initialize();
+  // denoises the first channel of the provided buffer,
+  // calculates the probability of Voice Activity in the buffer
+  // and stores it internally
+  void Denoise(AudioBuffer* audioBuffer, float noise_suppression_db);
+  // calculates RMS of the 1st channel in the given signal
+  // and stores it internally
+  void CalculateRMS(AudioBuffer* audioBuffer);
+  // encodes stored features into the given signal
+  void EncodeFeaturesToSignal(AudioBuffer* audioBuffer);
+
+  float ProcessDRREstimation(AudioBuffer* audioBuffer);
+  void SetDRREnabled(bool enabled);
+  void ResetDRR();
+
+  AudioFeatures* getCurrentFeatures();
+
+  VADState GetCurrentVADState() { return vad_state_; }
+
+  void SetAudioStatus(std::map<std::string, std::string> params);
+
+  float vad_threshold = kAroundDefaultDenoiserVADThreshold;
+
+ private:
+  // we need two resamplers to convert between rendering
+  // and rnnoise-specific sample rates
+  PushResampler<float> resampler_;
+  PushResampler<float> reverse_resampler_;
+  DenoiseState* denoise_state;
+  AudioFeatures audio_features = AudioFeatures{ 0.0, 0.0 };
+
+  VADState vad_state_ = Stopped;
+  int64_t vad_disable_timeout_ = 0;
+  int64_t muted_vad_disable_timeout_ = 0;
+  float vad_wema_ = 0.0f;
+  double drr_wema_ = 0.0;
+  bool is_microphone_on_ = true;
+  bool is_in_group_ = false;
+  bool drr_enabled_ = false;
+  double prev_kurtosis_ = 0.0;
+  double last_valid_residual_kurtosis_ = 0.0;
+  int last_valid_residual_kurtosis_distance_ = 0;
+  size_t drr_results_index_ = 0;
+  double* drr_results_ = NULL;
+  float hanning_window_[DRR_FRAME_SIZE];
+  RingBuffer* drr_buffer_ = NULL;
+  RingBuffer* running_avg_rms_buffer_ = NULL;
+  RingBuffer* running_avg_res_buffer_ = NULL;
+  FeatureExtractorSettings settings_ = FeatureExtractorSettings{
+    20,
+    0.6f,
+    0.1f,
+    0.0f,
+    0.2f,
+    5.2f,
+    6.0f,
+    0.0f,
+    1.5f,
+    true,
+    20.0f,
+    0.25f,
+    0.75f
+  };
+
+  void InitializeResamplersIfNeeded(int sample_rate_hz);
+
+  void ProcessVADState(float vad, int64_t now);
+  void ProcessMutedSpeakingState(float vad, int64_t now);
+  FullPipeline2Result FullPipelineV2(float *analysis_data, size_t analysis_data_len, float rms);
+  double FullPipelineV2Normalized(float *analysis_data, size_t analysis_data_len, float rms);
+
+  RTC_DISALLOW_COPY_AND_ASSIGN(AroundDenoiser);
+};
+
+} // namespace webrtc
+
+#endif
diff --git a/modules/audio_processing/audio_processing_impl.cc b/modules/audio_processing/audio_processing_impl.cc
index d7fbfb6569f67f087bca4660ac3eb807f55338be..e8eb1a01f7085bc9064459e20f0913e70e043784 100644
--- a/modules/audio_processing/audio_processing_impl.cc
+++ b/modules/audio_processing/audio_processing_impl.cc
@@ -39,6 +39,8 @@
 #include "system_wrappers/include/field_trial.h"
 #include "system_wrappers/include/metrics.h"
 
+#include "modules/audio_processing/around_biquad/mask_compensation_coefficients.h"
+
 #define RETURN_ON_ERR(expr) \
   do {                      \
     int err = (expr);       \
@@ -301,7 +303,9 @@ AudioProcessingImpl::AudioProcessingImpl(
   Initialize();
 }
 
-AudioProcessingImpl::~AudioProcessingImpl() = default;
+AudioProcessingImpl::~AudioProcessingImpl() {
+  TeardownDataChannelObserver();
+}
 
 int AudioProcessingImpl::Initialize() {
   // Run in a single-threaded manner during initialization.
@@ -415,12 +419,36 @@ void AudioProcessingImpl::InitializeLocked() {
   InitializePostProcessor();
   InitializePreProcessor();
   InitializeCaptureLevelsAdjuster();
+  InitializeAroundDenoiser();
+  InitializeAroundMaskCompensation();
+  InitializeAroundAgc();
+
+  if (config_.around_audio_processing.enabled) {
+    SetupDataChannelObserver();
+  }
 
   if (aec_dump_) {
     aec_dump_->WriteInitMessage(formats_.api_format, rtc::TimeUTCMillis());
   }
 }
 
+void AudioProcessingImpl::Stop() {
+  DataChannelController* data_channel_controller = DataChannelController::GetSharedDataChannelController();
+  if (data_channel_controller != NULL) {
+    AroundDenoiser::VADState vad_state = submodules_.around_denoiser->GetCurrentVADState();
+
+    if (vad_state == AroundDenoiser::Started || vad_state == AroundDenoiser::Sustained) {
+      data_channel_controller->SendAroundMessage("vad-stopped");
+    }
+
+    if (vad_state == AroundDenoiser::MutedStarted) {
+      data_channel_controller->SendAroundMessage("speaking-muted-stopped");
+    }
+  }
+
+  TeardownDataChannelObserver();
+}
+
 int AudioProcessingImpl::InitializeLocked(const ProcessingConfig& config) {
   UpdateActiveSubmoduleStates();
 
@@ -1194,7 +1222,9 @@ int AudioProcessingImpl::ProcessCaptureStreamLocked() {
     // capture signal saturation.
     // TODO(peah): Look into ensuring that this kind of tampering with the
     // AudioBuffer functionality should not be needed.
-    capture_buffer->set_num_channels(1);
+    // @AROUND: we disable this because the 2nd channel (if available) will
+    //          contain VAD and/or other metadata
+    // capture_buffer->set_num_channels(1);
   }
 
   if (submodules_.high_pass_filter &&
@@ -1267,6 +1297,10 @@ int AudioProcessingImpl::ProcessCaptureStreamLocked() {
     }
   }
 
+  if (config_.around_audio_processing.enabled) {
+    submodules_.around_denoiser->CalculateRMS(capture_buffer);
+  }
+
   if (submodules_.gain_control) {
     // TODO(peah): Add reporting from AEC3 whether there is echo.
     RETURN_ON_ERR(submodules_.gain_control->ProcessCaptureAudio(
@@ -1410,10 +1444,80 @@ int AudioProcessingImpl::ProcessCaptureStreamLocked() {
   }
   capture_.capture_output_used_last_frame = capture_.capture_output_used;
 
+  // @AROUND: Denoiser
+  if (config_.around_audio_processing.enabled) {
+    // Add mask compensation filter if enabled.
+    if (config_.around_audio_processing.mask_compensation) {
+      submodules_.around_mask_compensation->Process(capture_buffer);
+    }
+
+    // Always call this since user-VAD is computed within.
+    submodules_.around_denoiser->Denoise(capture_buffer, config_.around_audio_processing.noise_suppression_db);
+
+    auto features = submodules_.around_denoiser->getCurrentFeatures();
+
+    // Apply final digital AGC
+    submodules_.around_agc->Process(capture_buffer, features->vad_probability);
+
+    submodules_.around_denoiser->EncodeFeaturesToSignal(capture_buffer);
+
+    float drr_to_emit = submodules_.around_denoiser->ProcessDRREstimation(capture_buffer);
+
+    AroundDenoiser::VADState vad_state = submodules_.around_denoiser->GetCurrentVADState();
+    if (vad_state != previous_vad_state_) {
+      if (vad_state == AroundDenoiser::Stopped) {
+        submodules_.around_denoiser->ResetDRR();
+      }
+
+      DataChannelController* data_channel_controller = DataChannelController::GetSharedDataChannelController();
+      if (data_channel_controller != NULL) {
+        std::string vad_message("vad-stopped");
+        if (vad_state == AroundDenoiser::Started) vad_message = "vad-started";
+        if (vad_state == AroundDenoiser::Sustained) vad_message = "vad-sustained";
+        if (vad_state == AroundDenoiser::MutedStarted) vad_message = "speaking-muted-started";
+        if (vad_state == AroundDenoiser::MutedStopped) vad_message = "speaking-muted-stopped";
+
+        data_channel_controller->SendAroundMessage(vad_message);
+        previous_vad_state_ = vad_state;
+      }
+    }
+
+    if (drr_to_emit != 0.0f) {
+      DataChannelController* data_channel_controller = DataChannelController::GetSharedDataChannelController();
+      if (data_channel_controller != NULL) {
+        std::string drr_state = "drr-data;" + std::to_string(drr_to_emit);
+        data_channel_controller->SendAroundMessage(drr_state);
+      }
+    }
+  }
+
   capture_.was_stream_delay_set = false;
   return kNoError;
 }
 
+void AudioProcessingImpl::InitializeAroundDenoiser() {
+  // Create and initialize only if enabled and the Around NS doesn't already exist.
+  if (!submodules_.around_denoiser) {
+    submodules_.around_denoiser.reset(new AroundDenoiser());
+    submodules_.around_denoiser->Initialize();
+  }
+}
+
+void AudioProcessingImpl::InitializeAroundMaskCompensation() {
+  if (!submodules_.around_mask_compensation) {
+    submodules_.around_mask_compensation.reset(new AroundBiquad());
+  }
+  submodules_.around_mask_compensation->Initialize(MASK_COMPENSATION_COEFFS);
+}
+
+void AudioProcessingImpl::InitializeAroundAgc() {
+  if (!submodules_.around_agc) {
+    submodules_.around_agc.reset(new AroundAgc());
+  }
+  submodules_.around_agc->Initialize(proc_fullband_sample_rate_hz());
+  submodules_.around_agc->Reset();
+}
+
 int AudioProcessingImpl::AnalyzeReverseStream(
     const float* const* data,
     const StreamConfig& reverse_config) {
@@ -2207,4 +2311,36 @@ void AudioProcessingImpl::ApmStatsReporter::UpdateStatistics(
   static_cast<void>(stats_message_passed);
 }
 
+void AudioProcessingImpl::OnCommandReceived(std::string command, std::map<std::string, std::string> params) {
+  if (command == "audio_status") {
+    submodules_.around_denoiser->SetAudioStatus(params);
+  }
+
+  if (command == "noise_suppression") {
+    absl::optional<float> value = rtc::StringToNumber<float>(params["level"]);
+    if (!value.has_value()) {
+      RTC_LOG(LS_ERROR) << "AROUND ERROR: Invalid noise suppression level: " << params["level"];
+      return;
+    }
+
+    config_.around_audio_processing.noise_suppression_db = value.value();
+  }
+
+  if (command == "mask_compensation") {
+    config_.around_audio_processing.mask_compensation =
+      rtc::StringToNumber<unsigned int>(params["enabled"])
+      .value_or(config_.around_audio_processing.mask_compensation);
+
+    // If we just disabled, reset.
+    if (submodules_.around_mask_compensation &&
+        config_.around_audio_processing.mask_compensation) {
+      submodules_.around_mask_compensation->Reset();
+    }
+  }
+
+  if (command == "enable_drr") {
+    submodules_.around_denoiser->SetDRREnabled(params["enabled"] == "true");
+  }
+}
+
 }  // namespace webrtc
diff --git a/modules/audio_processing/audio_processing_impl.h b/modules/audio_processing/audio_processing_impl.h
index 7c541736589ca997ec4454abf6ebf72273d31548..91434ef3c61e9e5ea363acb4b8a0d7b9cacfdb46 100644
--- a/modules/audio_processing/audio_processing_impl.h
+++ b/modules/audio_processing/audio_processing_impl.h
@@ -22,6 +22,9 @@
 #include "modules/audio_processing/aec3/echo_canceller3.h"
 #include "modules/audio_processing/agc/agc_manager_direct.h"
 #include "modules/audio_processing/agc/gain_control.h"
+#include "modules/audio_processing/around_denoise/around_denoise.h"
+#include "modules/audio_processing/around_biquad/around_biquad.h"
+#include "modules/audio_processing/around_agc/around_agc.h"
 #include "modules/audio_processing/audio_buffer.h"
 #include "modules/audio_processing/capture_levels_adjuster/capture_levels_adjuster.h"
 #include "modules/audio_processing/echo_control_mobile_impl.h"
@@ -40,6 +43,7 @@
 #include "modules/audio_processing/rms_level.h"
 #include "modules/audio_processing/transient/transient_suppressor.h"
 #include "modules/audio_processing/voice_detection.h"
+#include "pc/data_channel_controller.h"
 #include "rtc_base/gtest_prod_util.h"
 #include "rtc_base/ignore_wundef.h"
 #include "rtc_base/swap_queue.h"
@@ -51,7 +55,8 @@ namespace webrtc {
 class ApmDataDumper;
 class AudioConverter;
 
-class AudioProcessingImpl : public AudioProcessing {
+class AudioProcessingImpl : public AudioProcessing,
+                            public AroundCommandObserver {
  public:
   // Methods forcing APM to run in a single-threaded manner.
   // Acquires both the render and capture locks.
@@ -64,6 +69,7 @@ class AudioProcessingImpl : public AudioProcessing {
                       std::unique_ptr<CustomAudioAnalyzer> capture_analyzer);
   ~AudioProcessingImpl() override;
   int Initialize() override;
+  void Stop() override;
   int Initialize(int capture_input_sample_rate_hz,
                  int capture_output_sample_rate_hz,
                  int render_sample_rate_hz,
@@ -138,6 +144,8 @@ class AudioProcessingImpl : public AudioProcessing {
 
   AudioProcessing::Config GetConfig() const override;
 
+  void OnCommandReceived(std::string command, std::map<std::string, std::string> params) override;
+
  protected:
   // Overridden in a mock.
   virtual void InitializeLocked()
@@ -281,6 +289,9 @@ class AudioProcessingImpl : public AudioProcessing {
       RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
   void InitializePostProcessor() RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
   void InitializeAnalyzer() RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
+  void InitializeAroundDenoiser() RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
+  void InitializeAroundMaskCompensation() RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
+  void InitializeAroundAgc() RTC_EXCLUSIVE_LOCKS_REQUIRED(mutex_capture_);
 
   // Initializations of render-only submodules, requiring the render lock
   // already acquired.
@@ -404,6 +415,9 @@ class AudioProcessingImpl : public AudioProcessing {
     std::unique_ptr<LevelEstimator> output_level_estimator;
     std::unique_ptr<VoiceDetection> voice_detector;
     std::unique_ptr<CaptureLevelsAdjuster> capture_levels_adjuster;
+    std::unique_ptr<AroundDenoiser> around_denoiser;
+    std::unique_ptr<AroundBiquad> around_mask_compensation;
+    std::unique_ptr<AroundAgc> around_agc;
   } submodules_;
 
   // State that is written to while holding both the render and capture locks
@@ -539,6 +553,8 @@ class AudioProcessingImpl : public AudioProcessing {
       agc_render_signal_queue_;
   std::unique_ptr<SwapQueue<std::vector<float>, RenderQueueItemVerifier<float>>>
       red_render_signal_queue_;
+
+  AroundDenoiser::VADState previous_vad_state_ = AroundDenoiser::Stopped;
 };
 
 }  // namespace webrtc
diff --git a/modules/audio_processing/gain_control_impl.cc b/modules/audio_processing/gain_control_impl.cc
index b5454c05ed4545b486bd0579991b6cbcd4c95505..df4b5a6caa14408ea92a998311703775eb269b37 100644
--- a/modules/audio_processing/gain_control_impl.cc
+++ b/modules/audio_processing/gain_control_impl.cc
@@ -149,7 +149,7 @@ void GainControlImpl::PackRenderAudioBuffer(
 int GainControlImpl::AnalyzeCaptureAudio(const AudioBuffer& audio) {
   RTC_DCHECK(num_proc_channels_);
   RTC_DCHECK_GE(AudioBuffer::kMaxSplitFrameLength, audio.num_frames_per_band());
-  RTC_DCHECK_EQ(audio.num_channels(), *num_proc_channels_);
+  RTC_DCHECK_GE(audio.num_channels(), *num_proc_channels_);
   RTC_DCHECK_LE(*num_proc_channels_, mono_agcs_.size());
 
   int16_t split_band_data[AudioBuffer::kMaxNumBands]
@@ -202,7 +202,7 @@ int GainControlImpl::ProcessCaptureAudio(AudioBuffer* audio,
   RTC_DCHECK(num_proc_channels_);
   RTC_DCHECK_GE(AudioBuffer::kMaxSplitFrameLength,
                 audio->num_frames_per_band());
-  RTC_DCHECK_EQ(audio->num_channels(), *num_proc_channels_);
+  RTC_DCHECK_GE(audio->num_channels(), *num_proc_channels_);
 
   stream_is_saturated_ = false;
   bool error_reported = false;
diff --git a/modules/audio_processing/high_pass_filter.cc b/modules/audio_processing/high_pass_filter.cc
index bff7209e96bedfb9b2e735abede7811caea7d0f7..debbd01a8405affad7dca2b5dac9bceabfccd2ae 100644
--- a/modules/audio_processing/high_pass_filter.cc
+++ b/modules/audio_processing/high_pass_filter.cc
@@ -66,9 +66,9 @@ HighPassFilter::~HighPassFilter() = default;
 
 void HighPassFilter::Process(AudioBuffer* audio, bool use_split_band_data) {
   RTC_DCHECK(audio);
-  RTC_DCHECK_EQ(filters_.size(), audio->num_channels());
+  RTC_DCHECK_LE(filters_.size(), audio->num_channels());
   if (use_split_band_data) {
-    for (size_t k = 0; k < audio->num_channels(); ++k) {
+    for (size_t k = 0; k < filters_.size(); ++k) {
       rtc::ArrayView<float> channel_data = rtc::ArrayView<float>(
           audio->split_bands(k)[0], audio->num_frames_per_band());
       filters_[k]->Process(channel_data);
diff --git a/modules/audio_processing/include/audio_processing.cc b/modules/audio_processing/include/audio_processing.cc
index f57ace7dbfadba938d650483ca33184df764e538..60348288085b9d31ddb1caee7e71296774c56377 100644
--- a/modules/audio_processing/include/audio_processing.cc
+++ b/modules/audio_processing/include/audio_processing.cc
@@ -212,6 +212,9 @@ std::string AudioProcessing::Config::ToString() const {
           << "}}, residual_echo_detector: { enabled: "
           << residual_echo_detector.enabled
           << " }, level_estimation: { enabled: " << level_estimation.enabled
+      << " }, around_audio_processing: { enabled: " << around_audio_processing.enabled
+      << ", noise_suppression_db: " << around_audio_processing.noise_suppression_db
+      << ", mask_compensation: " << around_audio_processing.mask_compensation
           << " }}";
   return builder.str();
 }
diff --git a/modules/audio_processing/include/audio_processing.h b/modules/audio_processing/include/audio_processing.h
index 590540132fb3fd4ed2d17cb3cea36b118742632e..fdcac19c4def40ed86bcf8678d9121ac3228ff55 100644
--- a/modules/audio_processing/include/audio_processing.h
+++ b/modules/audio_processing/include/audio_processing.h
@@ -58,9 +58,9 @@ class CustomProcessing;
 //
 // Must be provided through AudioProcessingBuilder().Create(config).
 #if defined(WEBRTC_CHROMIUM_BUILD)
-static constexpr int kAgcStartupMinVolume = 85;
+static constexpr int kAgcStartupMinVolume = 100;
 #else
-static constexpr int kAgcStartupMinVolume = 0;
+static constexpr int kAgcStartupMinVolume = 100;
 #endif  // defined(WEBRTC_CHROMIUM_BUILD)
 static constexpr int kClippedLevelMin = 70;
 
@@ -208,7 +208,7 @@ class RTC_EXPORT AudioProcessing : public rtc::RefCountInterface {
     } capture_level_adjustment;
 
     struct HighPassFilter {
-      bool enabled = false;
+      bool enabled = true;
       bool apply_in_full_band = true;
     } high_pass_filter;
 
@@ -227,6 +227,7 @@ class RTC_EXPORT AudioProcessing : public rtc::RefCountInterface {
       enum Level { kLow, kModerate, kHigh, kVeryHigh };
       Level level = kModerate;
       bool analyze_linear_aec_output_when_available = false;
+      bool encode_voice_activity = false;
     } noise_suppression;
 
     // Enables transient suppression.
@@ -383,7 +384,7 @@ class RTC_EXPORT AudioProcessing : public rtc::RefCountInterface {
     } gain_controller2;
 
     struct ResidualEchoDetector {
-      bool enabled = true;
+      bool enabled = false;
     } residual_echo_detector;
 
     // Enables reporting of `output_rms_dbfs` in webrtc::AudioProcessingStats.
@@ -391,6 +392,13 @@ class RTC_EXPORT AudioProcessing : public rtc::RefCountInterface {
       bool enabled = false;
     } level_estimation;
 
+    struct AroundAudioProcessing {
+      bool enabled = true;
+      float noise_suppression_db = 15.0f;
+      // Enables a high shelf boosting filter to correct for mask attenuation.
+      bool mask_compensation = false;
+    } around_audio_processing;
+
     std::string ToString() const;
   };
 
@@ -528,6 +536,8 @@ class RTC_EXPORT AudioProcessing : public rtc::RefCountInterface {
   // TODO(webrtc:5298): Change to return void.
   virtual int Initialize() = 0;
 
+  virtual void Stop() = 0;
+
   // The int16 interfaces require:
   //   - only `NativeRate`s be used
   //   - that the input, output and reverse rates must match
